src/sync/channel.zig

```diff
@@ -1,170 +1,426 @@
+            }
+        };
+
+        const Buffer = struct {
+            next: Atomic(?*Buffer),
+            slots: [BLOCK_CAP]Slot,
+
+            fn create(allocator: Allocator) !*Buffer {
+                const new = try allocator.create(Buffer);
+                @memset(&new.slots, Slot.uninit);
+                new.next = Atomic(?*Buffer).init(null);
+                return new;
+            }
+
+            fn destroy(block: *Buffer, start: usize, allocator: Allocator) void {
+                for (start..BLOCK_CAP - 1) |i| {
+                    const slot = &block.slots[i];
+
+                    if (slot.state.load(.acquire) & READ == 0 and
+                        slot.state.fetchOr(DESTROY, .acq_rel) & READ == 0)
+                    {
+                        return;
+                    }
+                }
+
+                allocator.destroy(block);
+            }
+
+            fn deinit(block: *Buffer, allocator: Allocator) void {
+                if (block.next.load(.monotonic)) |n| {
+                    n.deinit(allocator);
+                    allocator.destroy(n);
+                }
+            }
+        };
+
+        const Slot = struct {
+            value: T,
+            state: Atomic(usize),
+
+            const uninit: Slot = .{
+                .value = undefined,
+                .state = Atomic(usize).init(0),
+            };
+        };
+
+        pub fn init(allocator: Allocator, initial_capacity: usize) Self {
+            _ = initial_capacity; // TODO: do something with this
+            return .{
+                .head = Position.init(),
+                .tail = Position.init(),
                 .allocator = allocator,
             };
-            return self;
         }
 
-        pub fn deinit(self: *Self) void {
-            var buff_lock = self.buffer.lock();
-            var buff: *std.ArrayList(T) = buff_lock.mut();
-            buff.deinit();
-            buff_lock.unlock();
+        pub fn create(allocator: Allocator, initial_capacity: usize) !*Self {
+            const channel = try allocator.create(Self);
+            channel.* = Self.init(allocator, initial_capacity);
+            return channel;
+        }
 
-            self.allocator.destroy(self);
+        pub fn send(channel: *Self, value: T) !void {
+            var backoff = Backoff.init();
+            var tail = channel.tail.index.load(.acquire);
+            var block = channel.tail.block.load(.acquire);
+            var next_block: ?*Buffer = null;
+
+            while (true) {
+                const offset = (tail >> SHIFT) % LAP;
+                if (offset == BLOCK_CAP) {
+                    backoff.snooze();
+                    tail = channel.tail.index.load(.acquire);
+                    block = channel.tail.block.load(.acquire);
+                    continue;
+                }
+
+                if (offset + 1 == BLOCK_CAP and next_block == null) {
+                    next_block = try Buffer.create(channel.allocator);
```

Comment:
This allocation will be unnecessary if there is contention on the tail and another thread creates the next buffer and updates the tail before we do. Ideally we would only create this if we really know 100% that the current thread needs to provide the next block. But waiting on an allocation while holding an exclusive write lock seems even worse. So maybe this is the best way.

Reply:
I added a comment explaining why this allocation should happen before the CAS and not after.